File: ejemplos/general/02-web-scraper-simple/ejemplo_requests.py
"""
Ejemplo: Extraer múltiples elementos de una página con requests + BeautifulSoup

Qué enseña este ejemplo:
  - Cómo hacer una petición HTTP GET con requests.get()
  - Qué significa raise_for_status() y para qué sirve
  - Cómo crear un objeto BeautifulSoup (el "parser" del HTML)
  - Cómo buscar TODOS los elementos de un tipo con find_all()
  - Cómo manejar errores de red con try/except

Para tu reto (web scraper):
  Aplica el mismo patrón para extraer datos de books.toscrape.com
  (título + precio de varios libros), que es una página pública
  diseñada específicamente para practicar scraping.

Glosario de términos:
  requests     — librería para hacer peticiones HTTP (GET, POST, etc.)
  BeautifulSoup— librería para navegar y buscar dentro de HTML/XML
  html.parser  — el motor de Python que interpreta el texto HTML
                 (alternativa: lxml, más rápido pero requiere instalación)
  find_all()   — busca TODOS los elementos que coincidan con el selector
  find()       — busca solo el PRIMERO que coincida
  .text        — extrae el texto dentro de una etiqueta HTML (sin las etiquetas)
  .get("attr") — extrae el valor de un atributo HTML (ej: href, src, class)
"""

import requests
from bs4 import BeautifulSoup

# Esta URL es una página de ejemplo hecha para practicar scraping
URL = "https://quotes.toscrape.com/"


def obtener_frases():
    try:
        # 1. Hacer la petición GET
        respuesta = requests.get(URL, timeout=5)

        # 2. raise_for_status() lanza una excepción si el servidor devolvió
        #    un error (ej: 404 Not Found, 500 Server Error)
        #    Sin esto, requests no falla aunque la página no exista
        respuesta.raise_for_status()

        # 3. Crear el objeto BeautifulSoup para navegar el HTML
        #    respuesta.text es el HTML de la página como texto plano
        soup = BeautifulSoup(respuesta.text, "html.parser")

        # 4. Buscar TODOS los elementos con class="quote"
        #    find_all devuelve una lista, aunque no encuentre nada (lista vacía)
        elementos = soup.find_all("div", class_="quote")

        print(f"Se encontraron {len(elementos)} frases en la página.\n")

        for elemento in elementos:
            # .find() dentro de un elemento busca solo en ese bloque
            texto = elemento.find("span", class_="text").text
            autor = elemento.find("small", class_="author").text
            print(f'"{texto}"')
            print(f"  — {autor}\n")

    except requests.exceptions.ConnectionError:
        print("Error: No se pudo conectar. Revisa tu conexión a internet.")
    except requests.exceptions.Timeout:
        print("Error: La petición tardó demasiado (timeout).")
    except requests.exceptions.HTTPError as e:
        print(f"Error HTTP: {e}")


if __name__ == "__main__":
    obtener_frases()

File: ejemplos/general/02-web-scraper-simple/ejemplo_requests.rs
// Ejemplo: Extraer múltiples elementos de una página web en Rust con reqwest + scraper
//
// Qué enseña este ejemplo:
//   - Cómo hacer una petición HTTP GET asíncrona con reqwest
//   - Cómo extraer datos del HTML usando la librería scraper
//   - Cómo escribir datos estructurados en un archivo CSV
//
// Para tu reto (web scraper):
//   Dependencias necesarias en Cargo.toml:
//   reqwest = { version = "0.11", features = ["blocking"] }
//   scraper = "0.18"
//   csv = "1.3"
//
// Glosario de términos "raros":
//   reqwest::blocking::get — Hace la petición web en modo "bloqueante" (espera a que termine).
//   scraper::Html::parse_document — Convierte el texto HTML a una estructura navegable.
//   scraper::Selector::parse — Prepara una regla CSS (ej: ".quote") para buscar elementos.
//   unwrap() — Atajo en Rust para "si hay error, explota el programa". Útil en scripts rápidos.
//   csv::Writer — Crea un escritor que se encarga de formatear correctamente las comas y comillas en el CSV.

use scraper::{Html, Selector};
use csv::Writer;
use std::error::Error;

const URL: &str = "https://quotes.toscrape.com/";

fn main() -> Result<(), Box<dyn Error>> {
    // 1. Hacer la petición GET (bloqueante)
    let respuesta = reqwest::blocking::get(URL)?;
    
    // 2. Verificar que la respuesta sea OK y extraer el HTML como texto
    let html = respuesta.error_for_status()?.text()?;

    // 3. Crear el objeto Html navegable
    let documento = Html::parse_document(&html);

    // 4. Preparar los selectores CSS
    // Selector::parse compila la regla. El unwrap() asume que escribiste bien la regla CSS.
    let selector_quote = Selector::parse(".quote").unwrap();
    let selector_text = Selector::parse(".text").unwrap();
    let selector_author = Selector::parse(".author").unwrap();

    let elementos = documento.select(&selector_quote).collect::<Vec<_>>();
    println!("Se encontraron {} frases en la página.\n", elementos.len());

    // 5. Preparar el escritor CSV
    let mut escritor = Writer::from_path("frases.csv")?;
    escritor.write_record(&["texto", "autor"])?; // Cabeceras

    // 6. Iterar sobre cada frase y extraer datos
    for elemento in elementos {
        // Encontramos el primer .text dentro del bloque
        let texto = elemento.select(&selector_text)
            .next()
            .map(|t| t.text().collect::<String>())
            .unwrap_or_default();
            
        // Encontramos el primer .author dentro del bloque
        let autor = elemento.select(&selector_author)
            .next()
            .map(|a| a.text().collect::<String>())
            .unwrap_or_default();

        println!("\"{}\"", texto);
        println!("  — {}\n", autor);

        // Guardamos la fila en el CSV (la librería csv maneja las comas internas automáticamente)
        escritor.write_record(&[&texto, &autor])?;
    }

    escritor.flush()?; // Asegurar que todo se escriba al disco
    println!("✓ Datos guardados exitosamente en 'frases.csv'");

    Ok(())
}
